x-pack/plugins/ml/public/application/services/dashboard_service.ts

```diff
@@ -0,0 +1,113 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import { SavedObjectsClientContract } from 'kibana/public';
+import { htmlIdGenerator } from '@elastic/eui';
+import { useMemo } from 'react';
+import {
+  DASHBOARD_APP_URL_GENERATOR,
+  DashboardUrlGenerator,
+  SavedDashboardPanel,
+  SavedObjectDashboard,
+} from '../../../../../../src/plugins/dashboard/public';
+import { useMlKibana } from '../contexts/kibana';
+import { ViewMode } from '../../../../../../src/plugins/embeddable/public';
+
+export type DashboardService = ReturnType<typeof dashboardServiceProvider>;
+
+export function dashboardServiceProvider(
+  savedObjectClient: SavedObjectsClientContract,
+  kibanaVersion: string,
+  dashboardUrlGenerator: DashboardUrlGenerator
+) {
+  const generateId = htmlIdGenerator();
+
+  return {
+    /**
+     * Fetches dashboards
+     */
+    async fetchDashboards(query?: string) {
+      return await savedObjectClient.find<SavedObjectDashboard>({
+        type: 'dashboard',
+        perPage: 10,
+        search: query ? `${query}*` : '',
+        searchFields: ['title^3', 'description'],
+      });
+    },
+    /**
+     * Attaches embeddable panels to the dashboard
+     */
+    async attachPanels(
+      dashboardId: string,
+      dashboardAttributes: SavedObjectDashboard,
+      panelsData: Array<Pick<SavedDashboardPanel, 'title' | 'type' | 'embeddableConfig'>>
+    ) {
+      const panels = JSON.parse(dashboardAttributes.panelsJSON) as SavedDashboardPanel[];
+      const version = kibanaVersion;
+
+      for (const panelData of panelsData) {
+        const panelIndex = generateId();
+        const maxPanel =
+          panels.length > 0
+            ? panels.reduce((prev, current) =>
+                prev.gridData.y > current.gridData.y ? prev : current
+              )
+            : null;
+        panels.push({
+          panelIndex,
+          embeddableConfig: panelData.embeddableConfig as { [key: string]: any },
+          title: panelData.title,
+          type: panelData.type,
+          version,
+          gridData: {
+            h: 15,
+            i: panelIndex,
+            w: 24,
+            x: 0,
+            y: maxPanel ? maxPanel.gridData.y + maxPanel.gridData.h : 0,
```

Comment:
Could the addition of panels fill up dashboard rows, rather than adding them one per row vertically? I would prefer the two swimlanes I added here to go into one row:

![image](https://user-images.githubusercontent.com/7405507/84892405-705d4d80-b095-11ea-852e-df37b3bf4a7d.png)

Reply:
@mdefazio has raised a good alternative to this workflow, where the user selects to add both the Overall and 'View by' swimlanes in a single operation. Could we add both swimlanes into a single panel, to give the same layout as we have in the Anomaly Explorer? Appreciate this might need some rearchitecting, so maybe one to consider in a follow-up? Then if the user really wanted the overall and view by swimlanes in separate embeddables, they could add each individually in two separate 'Add to dashboard' operations.

Reply:
Done in https://github.com/elastic/kibana/pull/68784/commits/7d3e4e157ba345cfd4f1277acc4a0ca3a2a18b65